[BD-13] Deprecate ModuleSystem.render_template

pomegranited opened the pull request:
<!--
##
####         Note: the Lilac master branch has been created.  Please consider whether your change
    ####     should also be applied to Lilac.  If so, make another pull request against the
####         open-release/lilac.master branch, or ping @nedbat for help or questions.
##

Please give the pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply. You may link to information rather than copy it.
More details about the template are at https://github.com/edx/open-edx-proposals/pull/180
(link will be updated when that document merges)
-->

## Description

Builds on the shim added by https://github.com/edx/edx-platform/pull/28440 to deprecate the `ModuleSystem.render_template` attribute, in favor of using the added `MakoService.render_template`.

This change is only a refactoring, and should not affect Learners, Course Authors, or anyone else using edx-platform.

## Supporting information

* [Content Core Platform Simplication project](https://openedx.atlassian.net/wiki/spaces/AC/pages/3091629503/Content+Core+Platform+Simplification)
* [BLENDED-113](https://openedx.atlassian.net/browse/BLENDED-113) (edX internal ticket)
* [FAL-2304](https://tasks.opencraft.com/browse/FAL-2304) (OpenCraft internal ticket)

## Testing instructions

* LMS: https://pr28571.sandbox.opencraft.hosting/
* Studio: https://studio.pr28571.sandbox.opencraft.hosting/

LMS as learner:
1. Login as a demo learner (`honor@example.com` / `edx`)
2. Browse around the [BD-13 test course](https://discovery.pr28571.sandbox.opencraft.hosting/learning/course/course-v1:OpenCraft+BD-13+2021_1/home) and ensure that blocks render as expected, and that you can submit answers to problems. This course has all of the affected XBlocks in it, including an example Video block which shows the CC license wrapper (cf 7fd14678c2f3b4f772e3ae27b254082a34bfee7e).
3. Switch to the Legacy Course Experience and browse around in that too.

LMS as course staff:
1. Login as a demo staff user (`staff@example.com` / `edx`)
2. Navigate to a course and its Instructor Dashboard > Email tab
3. Ensure that you can see and use the HTML WYSIWYG editor to craft emails to students. (cf 96d92083748145aa41001d85a28f293b338239ff)

Studio:
1. Login as a demo staff (`staff@example.com` / `edx`)
2. Navigate to the [BD-13 test course](https://studio.pr28571.sandbox.opencraft.hosting/course/course-v1:OpenCraft+BD-13+2021_1) and browse through the blocks. Ensure they render and edit as expected.
3. Check that you can submit answers to problems, and save changes to blocks.
4. Preview unpublished changes and ensure they render as expected.
4. Publish changes, and ensure they render in the LMS as expected.

## Deadline

None

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-6000](https://openedx.atlassian.net/browse/OSPR-6000) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

symbolist commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -279,7 +279,7 @@ def student_view(self, context):  # lint-amnesty, pylint: disable=unused-argumen
         """
         fragment = Fragment()
-        fragment.add_content(self.system.render_template('word_cloud.html', {
+        fragment.add_content(self.system.descriptor_runtime.render_template('word_cloud.html', {
```
I am not clear why we need to explicitly specify `descriptor_system`. Shouldn't the attribute be gettable on `system` itself via the CombinedSystem [getattr](https://github.com/edx/edx-platform/blob/d6202880ae43845d1ce6f6027f7e10a5f48a034a/common/lib/xmodule/xmodule/x_module.py#L2015)?

symbolist replied:
Besides the `descriptor_runtime` will go away once we eliminate the `ModuleSystem`.

symbolist replied:
Another thing worth checking. Is `render_template()` really needed before the `ModuleSystem` is initialized? And if yes, for what? Maybe instead we should be removing the attribute in the `DescriptorSystem` and adding a service at the time the `ModuleSystem` is initialized.

pomegranited replied:
@symbolist Did you mean something like this? 1c8b5065f4009338343dd9a69f6f8520a9620f14

I can see adding the `replace_urls`, `replace_course_urls`, and `replace_jump_to_urls` functions to this service, and wrapping the provided `render_template` function to call these, so they don't have to be called by the XBlocks explicitly anymore.

symbolist replied:
Thanks. this looks like a more useful direction. A few comments:

Since the `render_template()` method is only for [rendering Mako templates](https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/common/djangoapps/edxmako/shortcuts.py#L157-L182) and the others can be used with any template engine, we should keep them separate. For `render_template()`, moving it to a `MakoService` in `common/djangoapps/edxmako/services.py` would be good. We can tackle the others later.

> I can see adding the replace_urls, replace_course_urls, and replace_jump_to_urls functions to this service, and wrapping the provided render_template function to call these, so they don't have to be called by the XBlocks explicitly anymore.

How many places is this happening? Will it also not result in the content being processed twice (the second time via `block_wrappers`) in places where `render_template()` is called in a `student_view()`?

Also did you check if the `render_template()` method in the `DescriptorSystem` is redundant? We can do this in a separate PR but we should also get rid of all the code in [mako_module.py](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/mako_module.py) since I do not think it is being used any more. In case `get_context()` is being used it can be copied over to the relevant XBlocks.

pomegranited replied:
@symbolist  Ok, e975541 does:

* move `render_template` to a new `MakoService`.
  And since [xblock `RuntimeShim`](https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/openedx/core/djangoapps/xblock/runtime/shims.py#L154) also deprecates `render_template` in favor of using the [`ResourceLoader.render_mako_template`](https://github.com/edx/xblock-utils/blob/0a8589acc99ecc608fb8fe4f6ad862eaf2bbc3ae/xblockutils/resources.py#L68-L76), I've deprecated this service call too, and will replace the platform calls to `runtime.render_template` with this preferred usage.
* reverted changes involving `replace_*urls`; we can deal with that separately.

I'll create a separate PR to remove DescriptorSystem.render_template and `MakoDescriptorSystem` -- I'm pretty sure these are not being used.

symbolist replied:
@pomegranited Thanks!

Hmm, can `ResourceLoader.render_mako_template()` be used for XBlocks that are in edx-platform? It does seem to look up the templates in the package versus `render_template()` which [looks at](https://github.com/edx/edx-platform/blob/0c164ad4c24d8ee2a5d5f0a35cd057f21581866d/cms/envs/common.py#L558) the directories specified in settings. If that is the case and we can move the templates around to make it work, that would be fantastic.

pomegranited replied:
Hmm.. I gave the `ResourceLoader` a try, and found there's a lot of dependent (and shared) templates that would need to be moved around too, e.g. `common/djangoapps/pipeline_mako/templates/static_content.html`, `cms/templates/widgets/metadata-only-edit.html`.

So will just stick with using the new MakoService instead.

symbolist commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -807,6 +792,27 @@ def rebind_noauth_module_to_user(module, real_user):
+                render_template=render_to_string,
+                # TODO (cpennington): This should be removed when all html from
+                # a module is coming through get_html and is therefore covered
+                # by the replace_static_urls code below
```
Hmm, it does not look like `render_to_string()` is doing the [processing](https://github.com/edx/edx-platform/blob/master/common/djangoapps/static_replace/__init__.py#L150) that is done in the next two methods? Maybe this comment is outdated?

pomegranited replied:
From what I can see, [`runtime.replace_*urls` is only being used by the CAPA ProblemBlock](https://github.com/edx/edx-platform/search?q=runtime.replace_urls); the rest of the XBlocks use them via the [block_wrappers](https://github.com/edx/edx-platform/blob/78c6be8beeb49ae6edb132147b6ecdaa21664f9c/lms/djangoapps/courseware/module_render.py#L705-L726) passed from the LMS.

So I think these could be removed from the ModuleSystem entirely, along with several other properties used only by CAPA modules (e.g. `cache`, `can_execute_unsafe_code`, `xqueue`, `get_python_lib_zip`). I'll do it as part of a separate task, but maybe these could be moved to the LoncapaSystem?

symbolist commented on common/djangoapps/edxmako/services.py:
```diff
@@ -0,0 +1,36 @@
+            'Use xblockutils.resources.ResourceLoader.render_mako_template or a JavaScript-based template instead.',
+            DeprecationWarning, stacklevel=2,
+        )
+        return self._render_template
```
Instead of passing in the `render_template` method, can we instead have a parameter `namespace_prefix=''`? And then instead of a property this can become:

```
def render_template(template_name, dictionary, namespace='main')
  return render_to_string(template_name, dictionary, namespace=self.namespace_prefix + namespace)
```

That way we don't need to pass in the whole code to the service but just the relevant parameter.

pomegranited replied:
Sure thing: https://github.com/edx/edx-platform/pull/28571/commits/ae32582477ca80905e4555e3f283027ff40351c5#diff-0f9960747639224845a14d42acee7d24eae92d50406e88a3142a7da45467c596R215

symbolist commented on cms/lib/xblock/authoring_mixin.py:
```diff
@@ -39,7 +43,7 @@ def visibility_view(self, _context=None):
         fragment = Fragment()
         from cms.djangoapps.contentstore.utils import reverse_course_url
-        fragment.add_content(self.system.render_template('visibility_editor.html', {
+        fragment.add_content(loader.render_mako_template('templates/visibility_editor.html', {
```
It doesn't look like this pattern is being used in `edx-platform`? So to keep the changeset small, let's just use the service for now. We can do a mass switch later if needed.

pomegranited replied:
Ok, will do.

pomegranited replied:
Sorry no, I need to keep this change to `AuthoringMixin` as-is. The issue isn't with the `AuthoringMixin` itself, it's with the tests. When the `AuthoringMixin` is used in Studio, it's using the `PreviewModuleSystem`, which does have the `MakoService`, and so everything works fine.

However, [test_authoring_mixin.py](https://github.com/open-craft/edx-platform/blob/jill/bd-13-render_template/cms/lib/xblock/test/test_authoring_mixin.py#L56-L60) uses the `CachingDescriptorSystem`, and since I reverted b8c177d9d30cae70698818ad0ae830da53263009, the descriptor system doesn't contain the `MakoService`, and so the tests throw a `NoSuchServiceError`.

However, I was able to revert the other similar change made by b8c177d9d30cae70698818ad0ae830da53263009  to the `PollBlock`: that can just use the `MakoService` and leave its template in place.

Does that make sense?

symbolist commented on common/lib/xmodule/xmodule/tests/test_editing_module.py:
```diff
@@ -22,8 +22,7 @@ class TabsEditingDescriptorTestCase(unittest.TestCase):
         super().setUp()
-        system = get_test_descriptor_system()
-        system.render_template = Mock(return_value="<div>Test Template HTML</div>")
+        system = get_test_descriptor_system(render_template=Mock(return_value="<div>Test Template HTML</div>"))
```
Is a custom mock needed here or will a default be fine?

pomegranited replied:
Ah yep! simple Mock() works fine here.

symbolist commented on common/lib/xmodule/xmodule/tests/test_xml_module.py:
```diff
@@ -347,8 +347,7 @@ def non_editable_metadata_fields(self):
 
-        system = get_test_descriptor_system()
-        system.render_template = Mock(return_value="<div>Test Template HTML</div>")
+        system = get_test_descriptor_system(render_template=Mock(return_value="<div>Test Template HTML</div>"))
```
Is a custom mock needed here or will a default be fine?

pomegranited commented on common/lib/xmodule/xmodule/studio_editable.py:
```diff
@@ -1,15 +1,21 @@
     This class is only intended to be used with an XBlock!
     """
+    _services_requested = {
+        'mako': 'need',
```
`StudioEditableBlock` needs the `mako` service now, but since it's not an XBlock, it doesn't have the `_services_requested` class property itself, so I borrowed this trick from [AuthoringMixin](https://github.com/edx/edx-platform/blob/3152c14192dcccac79b7b7a76eba91b380eba78a/cms/lib/xblock/authoring_mixin.py#L22-L24) to avoid this error incurred while initializing the decorated class:

```
AttributeError: type object 'StudioEditableBlock' has no attribute '_services_requested'
```

However, it can be an empty dict and perform the same duty.. I'll fix that.

pomegranited replied:
Added comment about this: https://github.com/edx/edx-platform/pull/28571/files#diff-37de9c229ad553ea6e2b8eb6b8551bd17dab0f41057d053a41c3f1dd6a791becR15-R17

pomegranited commented on common/lib/xmodule/xmodule/mako_module.py:
```diff
@@ -9,9 +9,15 @@
 
+        # Import is placed here to avoid model import at project startup.
+        from common.djangoapps.edxmako.services import MakoService
+        self._services['mako'] = MakoService()
```
I started to do that, but instead, moved this hack to the Instructor Dashboard where it's using the HtmlBlock, which [provoked this change in the first place](https://github.com/edx/edx-platform/pull/28571/commits/96d92083748145aa41001d85a28f293b338239ff).

pomegranited replied:
See https://github.com/open-craft/edx-platform/blob/jill/bd-13-render_template/lms/djangoapps/instructor/views/instructor_dashboard.py#L645-L666

pomegranited replied:
Ugh.. removing this caused other test failures:

* https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/35104/testReport/junit/cms.djangoapps.contentstore.tests.test_contentstore/MiscCourseTests/Run_Tests___cms_unit___test_module_preview_in_whitelist/
* https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/35104/testReport/junit/xmodule.xmodule.tests.test_library_root/TestLibraryRoot/Run_Tests___commonlib_unit___test_library_author_view/
* https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/35104/testReport/junit/xmodule.xmodule.tests.test_library_root/TestLibraryRoot/Run_Tests___commonlib_unit___test_library_author_view/

So I'm inclined to leave it in place, with a comment to explain why.

When we factor out MakoDescriptorSystem, then it can go away.

pomegranited commented on common/lib/xmodule/xmodule/word_cloud_module.py:
```diff
@@ -304,7 +305,7 @@ def studio_view(self, _context):
         """
         fragment = Fragment(
-            self.system.render_template(self.mako_template, self.get_context())
+            self.runtime.service('mako').render_template(self.mako_template, self.get_context())
```
Fixed: https://github.com/edx/edx-platform/pull/28571/files#diff-d67dfa1a671138a97d2e7b05266e9764d3a46fb721ec6a71f4cde3f8c273db59R283

pomegranited commented on openedx/core/lib/license/wrapper.py:
```diff
@@ -3,12 +3,12 @@
 
 
-def wrap_with_license(block, view, frag, context):  # pylint: disable=unused-argument
+def wrap_with_license(block, view, frag, context, mako_service=None):  # pylint: disable=unused-argument
```
Fixed: https://github.com/edx/edx-platform/pull/28571/files#diff-47add6baad5a51b3fae67c89253812925f6818463a67a55412082f53b26d8a04R6

pomegranited commented on common/djangoapps/edxmako/services.py:
```diff
@@ -0,0 +1,29 @@
+    A service for rendering XBlocks to HTML using mako templates.
+
+    Args:
+        render_template(function): function that renders the given context to the given template.
```
Fixed: https://github.com/open-craft/edx-platform/blob/jill/bd-13-render_template/common/djangoapps/edxmako/services.py#L15-L16

pomegranited commented on common/djangoapps/edxmako/tests.py:
```diff
@@ -208,3 +209,28 @@ def test_render_to_string_when_no_global_context_cms(self):
+
+
+def mock_render_to_string(template, context, namespace='main'):
+    return f"<template namespace={namespace}>{template} {context}</template>"
```
Yep! see 02621344c0453720b67ba23d453f59b03ff686c4

https://github.com/open-craft/edx-platform/blob/jill/bd-13-render_template/common/djangoapps/edxmako/tests.py#L214-L229

pomegranited commented:
@symbolist There's 1 error reported in the JS test, but it's unrelated to this work:

```
  CMS.Views.Metadata.VideoList
    ✓ Initialize
    Render
      ✓ is rendered in correct way
      ✓ is rendered with opened extra videos bar
      ✗ is rendered without opened extra videos bar
	Expected spy closeExtraVideosBar to have been called.
	@/home/runner/work/edx-platform/edx-platform/cms/static/js/spec/video/transcripts/videolist_spec.js:9:12828
	then/</</<@/home/runner/work/edx-platform/edx-platform/cms/static/common/js/vendor/jquery.js:3366:30
	fire@/home/runner/work/edx-platform/edx-platform/cms/static/common/js/vendor/jquery.js:3187:11
	fireWith@/home/runner/work/edx-platform/edx-platform/cms/static/common/js/vendor/jquery.js:3317:7

Firefox 61.0.0 (Linux 0.0.0) CMS.Views.Metadata.VideoList Render is rendered without opened extra videos bar FAILED
	Expected spy closeExtraVideosBar to have been called.
	@/home/runner/work/edx-platform/edx-platform/cms/static/js/spec/video/transcripts/videolist_spec.js:9:12828
	then/</</<@/home/runner/work/edx-platform/edx-platform/cms/static/common/js/vendor/jquery.js:3366:30
	fire@/home/runner/work/edx-platform/edx-platform/cms/static/common/js/vendor/jquery.js:3187:11
	fireWith@/home/runner/work/edx-platform/edx-platform/cms/static/common/js/vendor/jquery.js:3317:7
```

Does this need to be resolved before merging?

symbolist commented:
@ormsbee This is ready to go!

ormsbee commented:
@pomegranited: I realize this is entirely my fault for sitting on this PR too long, but may I please ask for a rebase? I should be able to merge tomorrow morning.

Thank you.

pomegranited commented:
@ormsbee  No worries at all! Those conflicts are all from my other https://github.com/edx/edx-platform/pull/29190 that you just merged, so rebasing would have been a necessity either way.

Fixed the conflicts and rebased, and the tests look good again, so this should be ready to merge now. Thank you! 😄

edx-webhook commented:
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/28571&entry.1484655318=https://github.com/pomegranited&entry.752974735=2021-08-29+10:33&entry.1917517419=2021-11-09+17:21&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

dianakhuang commented:
@pomegranited We were seeing issues with visibility settings for ORA and LTI blocks and the error "Error: service 'mako' was not requested", so we are going to revert this PR.

pomegranited commented:
Thanks for the heads up @dianakhuang and @jnlapierre -- if you could provide some details here, I'll address them with a follow-up PR.

dianakhuang commented:
@pomegranited

The steps to reproduce are this:
1. Create an LTI or ORA component in Studio
2. Click the 'gear' icon: ![Visibility Setting](https://user-images.githubusercontent.com/2952947/142267068-c48e140a-e19e-4b29-ba74-14348356c250.png)
3. We expect to see visibility settings, but instead we get the error message: 
```
Students will not be able to access this component. Re-edit your component to fix the error.

Error: Service 'mako' was not requested.
```
![Screenshot 2021-11-16 at 13 58 23](https://user-images.githubusercontent.com/2952947/142267206-707ef832-9418-4450-8761-49e6c8ad873b.png)

I did some digging on our own services, but I wasn't able to find the corresponding stacktrace in our Splunk logs.

pomegranited commented:
Thanks for the info @dianakhuang ! I can reproduce this on my devstack with ORA2, so will sort that one out.

But I was unable to reproduce the issue with an LTI component, and (other than `XBlock`) the `LTIBlock` doesn't seem to share any common base classes with `OpenAssessmentBlock`. Would it be possible for you to provide a course export which demonstrates this issue for LTI?

CC @jnlapierre

pomegranited commented:
@dianakhuang CC @jnlapierre I've identified the issue with ORA2, which was caused by a missed change to the AuthoringMixin, and would affect any XBlock which didn't explicitly `@need("mako")`. The [LTIBlock did `@need("mako")`](https://github.com/edx/edx-platform/pull/28571/files#), so I'm still not clear on why this would cause a problem -- maybe there's a different LTI XBlock out there somewhere in a separate repo?

But I'm re-submitting this PR, and will add a test to show the issue, then add the fix to show the issue has been resolved.
